// components/App.js
import React, {Component} from 'react'
import {connect} from 'react-redux'
import {fetchTable, handleScroll} from 'actions'

import Table from './Table'
import Pages from './Pages'
import TestButtons from './TestButtons'
import Search from './Search'
import Description from './Description'

class App extends Component {

	componentDidMount() {
		this.props.fetchTable()
		document.addEventListener('scroll', this.props.handleScroll);
	}

	componentWillUnmount() {
		document.removeEventListener('scroll', this.props.handleScroll);
	}

	render() {
		const {fetching, fetchFailure, selectTr} = this.props
		return (
			<div className='container'>	
				<TestButtons />
				<div className='row'>
					<div className='col-md-6'>
						<Search />
					</div>
					<div className='col-md-6'>
						<Pages />
					</div>
				</div>	
				<div className='row'>
					<div className='col-12'>
						{fetchFailure && <div className='failure-text'>Сорян данных нет</div>} 
						{fetching && <div className='preloader'></div>}
						<Table />
						{selectTr && <Description />}
					</div>
				</div>
			</div>
		)
	}
}

const mapStateToProps = state => ({
	fetchFailure: state.tablePage.failure,
	fetching: state.tablePage.fetching,
	selectTr: state.tablePage.select,
})

const mapDispatchToProps = {
	fetchTable,
	handleScroll,
}

export default connect(mapStateToProps, mapDispatchToProps)(App)
